File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/events/GiftCardTopUpRequested.java
package it.giannibombelli.wsc2026.giftcard.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

public record GiftCardTopUpRequested(
    GiftCardId aggregateId,
    Money requestedAmount,
    Money currentBalance
) implements GiftCardEvent {
    public GiftCardTopUpRequested {
        Require.requireArgument(aggregateId, "giftCardId");
        Require.requireArgument(requestedAmount, "requestedAmount");
        Require.requireArgument(currentBalance, "currentBalance");
    }
}

File: typescript/src/giftcard/domain/events/giftCardTopUpRequested.ts
// Evento emesso quando viene richiesta una ricarica della gift card.

import type { Event } from '@/common/domain/model/event.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { GiftCardId } from '../giftcard/giftCardId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type GiftCardTopUpRequested = Event<'GiftCardTopUpRequested', GiftCardId> & {
  readonly requestedAmount: Money;
  readonly currentBalance: Money;
};

export function giftCardTopUpRequested(
  aggregateId: GiftCardId,
  requestedAmount: Money,
  currentBalance: Money
): GiftCardTopUpRequested {
  requireArgument(aggregateId, 'giftCardId');
  requireArgument(requestedAmount, 'requestedAmount');
  requireArgument(currentBalance, 'currentBalance');
  return { kind: 'GiftCardTopUpRequested', aggregateId, requestedAmount, currentBalance };
}
